### PR TITLE
[SPARK-13919] [SQL] [WIP] Resolving the Conflicts of ColumnPruning and PushPredicateThroughProject

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

This PR is a follow-up of https://github.com/apache/spark/pull/11682. 

Now, `ColumnPruning` and `PushPredicateThroughProject` reverse each other's effect. Although it will not cause the max iteration now, some queries are not optimized to the best. 

For example, in the following query, 

``` scala
    val input = LocalRelation('a.int, 'b.string, 'c.double, 'd.int)
    val originalQuery =
      input.select('a, 'b, 'c, 'd,
        WindowExpression(
          AggregateExpression(Count('b), Complete, isDistinct = false),
          WindowSpecDefinition( 'a :: Nil,
            SortOrder('b, Ascending) :: Nil,
            UnspecifiedFrame)).as('window)).where('window > 1).select('a, 'c)
```

After multiple iteration of two rules of `ColumnPruning` and `PushPredicateThroughProject`, the optimized plan we generated is like:

```
Project [a#0,c#0]                                                                                                                                                                    
+- Filter (window#0L > cast(1 as bigint))                                                                                                                                            
   +- Project [a#0,c#0,window#0L]                                                                                                                                                    
      +- Window [(count(b#0),mode=Complete,isDistinct=false) windowspecdefinition(a#0, b#0 ASC, RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS window#0L], [a#0], [b#0 ASC]   
         +- LocalRelation [a#0,b#0,c#0,d#0]                                                                                                                                          
```

However, the expected optimized plan should be like:

```
Project [a#0,c#0]
+- Filter (window#0L > cast(1 as bigint))
   +- Project [a#0,c#0,window#0L]
      +- Window [(count(b#0),mode=Complete,isDistinct=false) windowspecdefinition(a#0, b#0 ASC, RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS window#0L], [a#0], [b#0 ASC]
         +- Project [a#0,b#0,c#0]
            +- LocalRelation [a#0,b#0,c#0,d#0]                                                                                                                                          
```

The solution of this PR is to split the rule `ColumnPruning` into two parts: `PushProjectThroughFilter` and `ColumnPruning`. The new rule `PushProjectThroughPredicate` runs before starting `ColumnPruning`. This PR also moves the rule `PushPredicateThroughProject` before the rules `SetOperationPushDown` and `PushPredicateThroughJoin` to ensure all the predicates can be pushed before `PushProjectThroughFilter` reverses the effect of the rule `PushPredicateThroughProject`.
#### How was this patch tested?

The existing test cases already expose the problem, but we need to add more regression tests to ensure the future code changes will not break it. 

TODO: add more test cases.
Will submit another PR for stopping pushing `Project` through the other operators in `ColumnPruning` if it contains non-deterministic expressions.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -306,14 +311,28 @@ object SetOperationPushDown extends Rule[LogicalPlan] with PredicateHelper {
...
+ * Attempts to eliminate the reading of unneeded columns from the query plan
+ * by pushing Project through Filter.
  *
- *  - Inserting Projections beneath the following operators:
- *   - Aggregate
- *   - Generate
- *   - Project <- Join
- *   - LeftSemiJoin
+ * Note: This rule could reverse the effects of PushPredicateThroughProject.
+ * This rule should be run before ColumnPruning for ensuring that Project can be
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I'm a little against to depending on rules order too much, sometimes we have to as other solutions are way too complex, but for this issue, can we try to find a more general solution?

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Yeah, I have the same concern. This PR is just to resolve the conflicts based on the current infrastructure. 

In my opinion, in each batch, we need a few rule sets. The order of rule sets do not matter. In each rule set, the order of rules matters. However, this is a fundamental design change. @marmbrus @rxin might have a better idea in this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -306,14 +311,28 @@ object SetOperationPushDown extends Rule[LogicalPlan] with PredicateHelper {
...
+ * pushed as low as possible.
+ */
+object PushProjectThroughFilter extends Rule[LogicalPlan] {
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: We does not actual PUSH project through filter, we create new Project before to prune some columns.

As I said in another PR, we remove the those Project before filter.

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: @davies 

The naming of this rule is not right, but I still think this PR fixes the fundamental issue between the conflicts of `ColumnPruning` and `PushPredicateThroughProject`. If we do not take the ideas of this PR, I can find a test case to show the minor fix in `ColumnPruning` does not cover all the cases.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -91,6 +92,10 @@ abstract class Optimizer extends RuleExecutor[LogicalPlan] {
       SimplifyCasts,
       SimplifyCaseConversionExpressions,
       EliminateSerialization) ::
+    // Because ColumnPruning is called after PushPredicateThroughProject, the predicate push down
+    // is reversed. This batch is to ensure Filter is pushed below Project, if possible.
+    Batch("Push Predicate Through Project", Once,
+      PushPredicateThroughProject) ::
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Put this role in a separate batch is not correct, some other filter push down rules depend on this one.

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I did not remove it from the original batch. Just added the extra batch here.

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: oh, I missed that, sorry.

**Pull request closed without merging**